src/providers/wfs/qgswfsdataitems.cpp

```diff
@@ -192,6 +195,84 @@ void QgsWfsRootItem::newConnection()
 }
 #endif
 
+
+//////
+
+
+QgsDataItem *QgsWfsDataItemProvider::createDataItem( const QString &path, QgsDataItem *parentItem )
+{
+  QgsDebugMsg( "thePath = " + path );
+  if ( path.isEmpty() )
+  {
+    return new QgsWfsRootItem( parentItem, QStringLiteral( "WFS" ), QStringLiteral( "wfs:" ) );
+  }
+
+  // path schema: wfs:/connection name (used by OWS)
+  if ( path.startsWith( QLatin1String( "wfs:/" ) ) )
+  {
+    QString connectionName = path.split( '/' ).last();
+    if ( QgsWfsConnection::connectionList().contains( connectionName ) )
+    {
+      QgsWfsConnection connection( connectionName );
+      return new QgsWfsConnectionItem( parentItem, QStringLiteral( "WFS" ), path, connection.uri().uri() );
+    }
+  }
+  else if ( path.startsWith( QLatin1String( "geonode:/" ) ) )
+  {
+    QString connectionName = path.split( '/' ).last();
+    if ( QgsGeoNodeConnectionUtils::connectionList().contains( connectionName ) )
+    {
+      QgsGeoNodeConnection connection( connectionName );
+
+      QString url = connection.uri().param( "url" );
+      QgsGeoNodeRequest geonodeRequest( url, true );
+
+      QgsWFSDataSourceURI sourceUri( geonodeRequest.fetchServiceUrlsBlocking( QStringLiteral( "WFS" ) )[0] );
+
+      QgsDebugMsg( QString( "WFS full uri: '%1'." ).arg( QString( sourceUri.uri() ) ) );
+
+      return new QgsWfsConnectionItem( parentItem, QStringLiteral( "WFS" ), path, sourceUri.uri() );
+    }
+  }
+
+  return nullptr;
+}
+
+QVector<QgsDataItem *> QgsWfsDataItemProvider::createDataItems( const QString &path, QgsDataItem *parentItem )
+{
+  QVector<QgsDataItem *> items;
+  if ( path.startsWith( QLatin1String( "geonode:/" ) ) )
+  {
+    QString connectionName = path.split( '/' ).last();
+    if ( QgsGeoNodeConnectionUtils::connectionList().contains( connectionName ) )
+    {
+      QgsGeoNodeConnection connection( connectionName );
+
+      QString url = connection.uri().param( "url" );
+      QgsGeoNodeRequest geonodeRequest( url, true );
+
+      QStringList encodedUris( geonodeRequest.fetchServiceUrlsBlocking( QStringLiteral( "WFS" ) ) );
+
+      if ( !encodedUris.isEmpty() )
+      {
+        Q_FOREACH ( QString encodedUri, encodedUris )
```

Comment:
@nyalldawson  Do we prefer for range loops here?

Reply:
Fixed (Also optimised some map/hash based loops, where the keys where being iterated and then the values retrieved immediately. In this case it's much more efficient to iterate over the original container and avoid the multiple value lookups)